### PR TITLE
[ASAP] CP-47431: Use un-patched Json.NET in C# SDK and package SDK in releases. Update PS SDK to use NuGet reference to XenServer.NET

### PR DESCRIPTION
**Pull request by danilo-delbusso**
PR contains two main changes:
- CP-47431: Use un-patched Json.NET in C# SDK + minor changes to the source and NuGet package
- Improve "sanity check" run on every PR and push to ensure SDK is generated and C# SDK compiles and builds as expected.
- Also update PS SDK to use NuGet references and build for every PR and release
- Add an **unsigned** NuGet build to the release artefacts. Please see https://github.com/danilo-delbusso/xen-api/releases/tag/v99.2.0 for an example of the output. Also add PowerShell SDKs

To achieve this I split up the intialisation of a XenAPI environment in a separate composite action so it can be reused

### CODE REVIEW

File: .github/workflows/release.yml
Comment (psafont) [inline .github/workflows/release.yml:86]: why is this step duplicated? Doesn't look right

Comment (danilo-delbusso) [reply .github/workflows/release.yml:86]: good point we might be able to make it a single call, I see some things in actions/download-artifact that I can try 👀

Comment (danilo-delbusso) [reply .github/workflows/release.yml:86]: just noticed you meant the fact that the python files are being downloaded twice, not that I'm calling the action twice! 🤦

**Comment by danilo-delbusso** [comment]
rebased on master and added a commit based on PR comments

however, I am putting this one into draft since I noticed we need changes to the PS SDK

File: .github/workflows/setup-xapi-environment/action.yml
Comment (psafont) [inline .github/workflows/setup-xapi-environment/action.yml:59]: Can the cleanup step be added here somehow? (The unpin business)

Comment (danilo-delbusso) [reply .github/workflows/setup-xapi-environment/action.yml:59]: Good point I must've forgotten to add it again

Comment (danilo-delbusso) [reply .github/workflows/setup-xapi-environment/action.yml:59]: Looks like it's not supported just yet - https://github.com/actions/runner/issues/1478

Should I add it to `release.yml` manually though? We call this action from there but there's no cleanup

**Comment by danilo-delbusso** [comment]
Rebased on `master` to fix `Makefile` conflict and added 7e6a66cc0028a8e32034ed64709b5b8084a192b7 to address review commments

File: .github/workflows/setup-xapi-environment/action.yml
Comment (danilo-delbusso) [reply .github/workflows/setup-xapi-environment/action.yml:59]: done, in 7e6a66cc0028a8e32034ed64709b5b8084a192b7

File: ocaml/sdk-gen/csharp/templates/FriendlyErrorNames.Designer.mustache
Comment (kc284) [inline ocaml/sdk-gen/csharp/templates/FriendlyErrorNames.Designer.mustache:25]: The class and properties need to be public so implementing code can see them (remember that when we generated it with `resgen.exe` we passed `/publicClass XenAPI.FriendlyErrorNames.resources`).

File: ocaml/sdk-gen/csharp/FriendlyErrorNames.resx
Comment (kc284) [inline ocaml/sdk-gen/csharp/FriendlyErrorNames.resx:117]: This needs a change in Failure.cs (personally I'd like to remove those because they are xencenterisms and they rather belong there and not in the SDK, but it can wait for a future PR).

File: ocaml/sdk-gen/powershell/autogen/README_51.md
Comment (kc284) [inline ocaml/sdk-gen/powershell/autogen/README_51.md:77]: This is the README for 5.1, please restore the previous text because the paths are not the same as PS7.

File: ocaml/sdk-gen/powershell/autogen/README_51.md
Comment (kc284) [inline ocaml/sdk-gen/powershell/autogen/README_51.md:86]: Here and everywhere below: remove `7`, this is the 5.1 README.

File: ocaml/sdk-gen/powershell/autogen/README_51.md
Comment (kc284) [inline ocaml/sdk-gen/powershell/autogen/README_51.md:127]: Same here, please restore the paths for 5.1.

File: .github/workflows/setup-xapi-environment/action.yml
Comment (psafont) [inline .github/workflows/setup-xapi-environment/action.yml:20]: This step is not needed in all the the workflows, why not do the step explicitly instead?

Comment (danilo-delbusso) [reply .github/workflows/setup-xapi-environment/action.yml:20]: Mistake on my end. This is only needed for SDK gen, right?

I should move it to the reusable workflow instead of this action

File: .github/workflows/release.yml
Comment (psafont) [inline .github/workflows/release.yml:86]: There's quite a bit of repetition here, this would work as well:
```suggestion
gh release create ${{ github.ref_name }} --repo ${{ github.repository }} --generate-notes dist/* PowerShell-SDK-5.x-prerelease-unsigned.zip PowerShell-SDK-7.x-prerelease-unsigned.zip
```

If you want to have an artifact per line, maybe using `/` could work?

File: ocaml/sdk-gen/powershell/autogen/dune
Comment (kc284) [inline ocaml/sdk-gen/powershell/autogen/dune:2]: The copy to .txt was done as a courtesy to Windows users because the latter does not know what to do with files without extensions and needs an extra click to open the file. Personally I'm not too fussed, plus that it adds complexity, but might any users complain?

File: ocaml/sdk-gen/csharp/autogen/README.md
Comment (kc284) [inline ocaml/sdk-gen/csharp/autogen/README.md:19]: Since we're doing markdown, literal links should be in angle brackets.

File: .github/workflows/release.yml
Comment (danilo-delbusso) [reply .github/workflows/release.yml:86]: Well spotted. I had assumed (but not checked) that I couldn't list them like that!

**Review by kc284 — CHANGES_REQUESTED** [review]
First round of comments just by reading the PR (will also have a look at the generated code).

File: ocaml/sdk-gen/powershell/autogen/dune
Comment (danilo-delbusso) [reply ocaml/sdk-gen/powershell/autogen/dune:2]: my thought process was also that the PS SDK is now cross platform and so I followed the standard for no extenion in `LICENSE` (since Linux distros mostly assume it's a plain text file with no extension)

the real reason I changed this is that I like to follow the standard of `LICENSE` having no extension but honestly it's not that important.

If you prefer, I can revert this change, just let me know

**Comment by danilo-delbusso** [comment]
spotted an issue and had to force-push twice.

see this diff for the latest changes:

https://github.com/xapi-project/xen-api/compare/8e8b981e459d2893ee929b586dcb55a7ce5b8cb4..63d86e3fc7f79e31ede2c7a42f179a5aa032020

File: .github/workflows/setup-xapi-environment/action.yml
Comment (danilo-delbusso) [reply .github/workflows/setup-xapi-environment/action.yml]: guess it was needed for OCaml tests too, should I re-add it to the composite action or call it manually?

My idea behind `setup-xapi-environment` was to have one reusable action for any instance so if it's used by both workflows I'd add the wget to the action - thoughts?

**Comment by danilo-delbusso** [comment]
spotted an issue and had to force-push a couple times.

see this diff for the latest changes, to make it easier to review them:

https://github.com/xapi-project/xen-api/compare/8e8b981e459d2893ee929b586dcb55a7ce5b8cb4..ed1cafeacca924280529fabe97ed1f82805f5d73

File: ocaml/sdk-gen/powershell/autogen/dune
Comment (kc284) [reply ocaml/sdk-gen/powershell/autogen/dune:2]: I'm ok to leave it for now, we can always do it later if anyone complains.

File: ocaml/sdk-gen/csharp/templates/FriendlyErrorNames.Designer.mustache
Comment (kc284) [inline ocaml/sdk-gen/csharp/templates/FriendlyErrorNames.Designer.mustache:27]: I did say "the class and properties" in the previous comment - the accessibility of the fields and constructor does not need changing.

File: ocaml/sdk-gen/powershell/autogen/src/XenServerPowerShell.sln
Comment (kc284) [inline ocaml/sdk-gen/powershell/autogen/src/XenServerPowerShell.sln:1]: I forgot to ask earlier, why do we need a solution file for PS and not for C#? (Strictly speaking neither is essential).

File: ocaml/sdk-gen/powershell/autogen/README_51.md
Comment (kc284) [inline ocaml/sdk-gen/powershell/autogen/README_51.md:80]: I think the version might not be needed for the old powershell, but if you want to use it, let's write 5.1 to be more precise. Same below.

**Review by kc284 — CHANGES_REQUESTED** [review]

File: ocaml/sdk-gen/powershell/autogen/src/XenServerPowerShell.sln
Comment (danilo-delbusso) [reply ocaml/sdk-gen/powershell/autogen/src/XenServerPowerShell.sln:1]: so when building from VS22 it kept creating it. I know it's not necessary but I liked the idea that if one downloaded the source it'd just start without generating new files or modifying them.

I think it's been created as a result of switching to NuGet (though it technically isn't a requirement)

File: ocaml/sdk-gen/csharp/templates/FriendlyErrorNames.Designer.mustache
Comment (danilo-delbusso) [reply ocaml/sdk-gen/csharp/templates/FriendlyErrorNames.Designer.mustache]: you're right - sorry. I have gone ahead and taken the output of the `ResGen.exe` we use when building

File: ocaml/sdk-gen/powershell/autogen/src/XenServerPowerShell.sln
Comment (kc284) [reply ocaml/sdk-gen/powershell/autogen/src/XenServerPowerShell.sln:1]: Generally VS does ask one if they want to save the solution, but it does not create it if one says no. I believe the reason why we never provided solution files is that it is unknown what the users will do with the projects, i.e. modify/build them separately or include them in their own solution in which case the provided solution file is not needed. (Personally, I have my own solution with both C# and PS projects together, but this may not be necessarily what the majority of users do, hence I'm not suggesting shipping it). I think if we want to provide solution files we should do it either for both C# and PS or for none, otherwise it looks untidy, but I'd go for none, because it's another 2 files to maintain.

Comment (danilo-delbusso) [reply ocaml/sdk-gen/powershell/autogen/src/XenServerPowerShell.sln:1]: > I think if we want to provide solution files we should do it either for both C# and PS or for none, otherwise it looks untidy, but I'd go for none, because it's another 2 files to maintain.

You make a compelling point, I fully agree. I'll revert the addition of the `sln` ^^

**Review by psafont — APPROVED** [review]
The workflow changes seem fine to me

**Review by CitrixChris — APPROVED** [review]

**Review by kc284 — APPROVED** [review]

**Comment by kc284** [comment]
@danilo-delbusso I noticed the "don't merge" label was added today, is there anything new blocking this PR?

**Comment by danilo-delbusso** [comment]
> @danilo-delbusso I noticed the "don't merge" label was added today, is there anything new blocking this PR?

Nope, just wanted to make sure you had a chance to review it before it was merged :)

File: .github/workflows/main.yml
Comment (psafont) [inline .github/workflows/main.yml:171]: is there anything against
```suggestion
      xapi_version: ${{ env.XAPI_VERSION }}
```

Comment (danilo-delbusso) [reply .github/workflows/main.yml:171]: honestly just the fact that I don't want people thinking that the PR artefacts are valid for real usage. 

I know we also flag release ones as unsigned and prerelease but these are generated for every PR and I wanted to make it even clearer that they should not be used at all

Comment (psafont) [reply .github/workflows/main.yml:171]: XAPI_VERSION should be 0.0.0 most of the time. I'm just wary of having 2 sources of truth

Comment (danilo-delbusso) [reply .github/workflows/main.yml:171]: ah! that's a good point - I'll change it 👍

Comment (danilo-delbusso) [reply .github/workflows/main.yml]: done, see https://github.com/xapi-project/xen-api/compare/15e55edc3e51bfc8d3b9a88ea61b2da32052dba8..361e342b9c6668643c28c09bcb14571af0aa6540

**Pull request merged**
